### ISSUE TITLE
Is it legal to produce events from the background consumer thread using ProducerType.MULTI?

### ISSUE THREAD

**Issue by carterkozak** [LMAX-Exchange/disruptor#307]
I'm trying to debug an issue where the queue appears to be full, but my background thread is in `runnable` state spinning on `at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:190)` without making any progress. I'm curious if there's a memory safety issue when the consumer thread adds elements to the queue, but I don't see this mentioned in the documentation.

The issue is odd because the background thread never appears to do a great deal of processing, and the queue grows slowly over time as through the state has become lost.

Any information or ideas are greatly appreciated, thanks!

**Comment by ocoanet** [comment]
Your are queuing events from an event handler. I call this pattern self-enqueuing. Self-enqueuing is very dangerous because it can generate deadlocks when the ring buffer is full. You can read [this issue](https://github.com/disruptor-net/Disruptor-net/issues/41) about self-enqueuing in the Disruptor-net project.

**Comment by carterkozak** [comment]
Right, I think we should replace all self-enqueuing with synchronous operations, however this implementation does already use tryEnqueue and fall back to synchronous handling on the background thread when the queue is full. I'm not confident that removing the possibility of self-enqueueing would solve the problem in this case. The consumer thread is simply spinning through `BatchEventProcessor` without making progress on the queue, perhaps there's a memory ordering problem and we're in an unexpected state?

**Comment by ocoanet** [comment]
If your processor is busy spinning without making progress (with `availableSequence < nextSequence`) , my only guess would be that a publisher was able to acquire a sequence but did not publish it. It could be that the publisher is blocked or that the publisher failed to publish the sequence.

**Comment by mzheravin** [comment]
Have you tried different wait strategies?
Your 'background cosumer thread' is the tread initiated by disruptor, or you are trying to attach your thread?

**Comment by carterkozak** [comment]
> It could be that the publisher is blocked or that the publisher failed to publish the sequence.

We only publish via `RingBuffer.publishEvent` and `RingBuffer.tryPublishEvent`, without interacting with the sequence directly. It's possible something unexpected is occurring in our SequenceReportingEventHandler that prevents the sequence callback from being invoked, but that should be very unlikely (though this issue is also incredibly uncommon!).

> Have you tried different wait strategies?

I've seen this occur with both TimeoutBlockingWaitStrategy and BlockingWaitStrategy, unfortunately it has occurred about once every six months. I'd assume bad dimms, but we're running in AWS where everything is uses ECC.

> Your 'background cosumer thread' is the tread initiated by disruptor, or you are trying to attach your thread?

Yes, the thread created by disruptor.

**Comment by ocoanet** [comment]
Could you post your publication code and describe from what thread you are publishing?

**Comment by carterkozak** [comment]
Of course, this is using log4j2

Publication occurs here: https://github.com/apache/logging-log4j2/blob/70ce3183f3c811e3cde2052b60587e8a96d0a8ee/log4j-core/src/main/java/org/apache/logging/log4j/core/async/AsyncLogger.java#L228-L259

We're using the [MULTI](https://github.com/apache/logging-log4j2/blob/70ce3183f3c811e3cde2052b60587e8a96d0a8ee/log4j-core/src/main/java/org/apache/logging/log4j/core/async/AsyncLoggerDisruptor.java#L108-L109) producer type.
Publication can occur on any thread that logs, including the disruptor thread if the layout calls toString on an object which logs inside of a toString. There's special handling to avoid enqueueing from the disruptor thread when `tryPublish` fails, where events are recorded synchronously.

Publishes through https://github.com/apache/logging-log4j2/blob/release-2.x/log4j-core/src/main/java/org/apache/logging/log4j/core/async/AsyncLoggerDisruptor.java

EventHandler: https://github.com/apache/logging-log4j2/blob/release-2.x/log4j-core/src/main/java/org/apache/logging/log4j/core/async/RingBufferLogEventHandler.java

The disruptor thread is runnable:

```
"Log4j2-TF-1-AsyncLogger[sls-service]-1" #24 daemon prio=5 os_prio=0 cpu=7707684.49ms elapsed=23625.18s tid=0x00007fd28782a000 nid=0x45c2 runnable  [0x00007fc56066d000]
   java.lang.Thread.State: RUNNABLE
        at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:190)
        at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
        at java.lang.Thread.run(java.base@11.0.8/Thread.java:834)
```

A producer thread is parked waiting to move forward, stuck:

```
"BackgroundWriteQueueProcessor RUNNING" #356 prio=5 os_prio=0 cpu=612194.80ms elapsed=23615.36s tid=0x00007fc559b7d000 nid=0x4753 runnable  [
0x00007fc525f81000]
   java.lang.Thread.State: TIMED_WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@11.0.8/Native Method)
        at java.util.concurrent.locks.LockSupport.parkNanos(java.base@11.0.8/LockSupport.java:357)
        at com.lmax.disruptor.MultiProducerSequencer.next(MultiProducerSequencer.java:136)
        at com.lmax.disruptor.MultiProducerSequencer.next(MultiProducerSequencer.java:105)
        at com.lmax.disruptor.RingBuffer.publishEvent(RingBuffer.java:465)
        at com.lmax.disruptor.dsl.Disruptor.publishEvent(Disruptor.java:331)
        at org.apache.logging.log4j.core.async.AsyncLoggerDisruptor.enqueueLogMessageWhenQueueFull(AsyncLoggerDisruptor.java:230)
        - locked <0x00007fc5fa1eb278> (a java.lang.Object)
        at org.apache.logging.log4j.core.async.AsyncLogger.handleRingBufferFull(AsyncLogger.java:246)
        at org.apache.logging.log4j.core.async.AsyncLogger.publish(AsyncLogger.java:230)
        at org.apache.logging.log4j.core.async.AsyncLogger.logWithThreadLocalTranslator(AsyncLogger.java:225)
        at org.apache.logging.log4j.core.async.AsyncLogger.access$000(AsyncLogger.java:67)
        at org.apache.logging.log4j.core.async.AsyncLogger$1.log(AsyncLogger.java:152)
        at org.apache.logging.log4j.core.async.AsyncLogger.log(AsyncLogger.java:136)
        at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2198)
        at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(AbstractLogger.java:2152)
        at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2135)
        at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2016)
        at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1875)
        at org.apache.logging.slf4j.Log4jLogger.info(Log4jLogger.java:179)
<application code which logs>
```

Other threads are blocked waiting on the `0x00007fc5fa1eb278` lock, which we introduced to log4j to avoid causing all blocked threads to spin when blocking on a full queue (we saw 100% cpu utilization across all cores if the queue ever managed to fill).

**Comment by carterkozak** [comment]
This occurred most recently using zulu11.41_23-ca (11.0.8) using g1gc with string deduplication: `-XX:+UseG1GC -XX:+UseStringDeduplication`
Interestingly when the queue began to grow and the handler stopped processing, there was only ~1 event/second being produced.

**Comment by ocoanet** [comment]
> I'm trying to debug an issue where the queue appears to be full
> [...]
> Interestingly when the queue began to grow and the handler stopped processing, there was only ~1 event/second being produced.

Ok so the fact that the ring buffer is full is only a consequence of the stopped handler?

Was the `BatchEventProcessor` blocked in `SequenceBarrier.waitFor` or was it blocked because `availableSequence < nextSequence`?

**Comment by carterkozak** [comment]
Metrics reported no space available on the queue, so it’s blocked on ‘availableSequence < nextSequence’

**Comment by ocoanet** [comment]
If the issue was a self-enqueue deadlock, it would be visible in the `BatchEventProcessor` call stack. But this is not the case here. There is simply a blocked publisher thread caused by a full ring buffer.

The only time I encountered a hung `BatchEventProcessor` in my applications, the problem was a bug in the publication code where an entry publication was skipped. But because you always use `EventTranslator`, this is should not the case here either.

It am out of ideas here, sorry. I think we need help from the maintainers 🙂

**Comment by ocoanet** [comment]
On a side note, I am not sure to like this code:
```java
    boolean tryPublish(final RingBufferLogEventTranslator translator) {
        try {
            // Note: we deliberately access the volatile disruptor field afresh here.
            // Avoiding this and using an older reference could result in adding a log event to the disruptor after it
            // was shut down, which could cause the publishEvent method to hang and never return.
            return disruptor.getRingBuffer().tryPublishEvent(translator);
        } catch (final NullPointerException npe) {
            // LOG4J2-639: catch NPE if disruptor field was set to null in stop()
            logWarningOnNpeFromDisruptorPublish(translator);
            return false;
        }
    }
```
Even with a volatile `disruptor`, the field must be loaded before invoking the method, and the `disruptor` could be shut down after the load but before invoking `getRingBuffer`. Also, are you sure that your translator will never raise `NullPointerException`?

**Comment by carterkozak** [comment]
> It am out of ideas here, sorry. I think we need help from the maintainers slightly_smiling_face

No worries, I appreciate your help :-)

> On a side note, I am not sure to like this code

You're right, that is strange. I'll analyze if I can update that to a single volatile read and an explicit null check. I don't expect the translator to throw a NPE, but even if it did I think the RingBuffer `sequencer.publish(sequence);` would be invoked. We might end up publishing an event twice (first via tryPublish, then following the blocking path), but it shouldn't produce risk beyond that.

**Comment by grumpyjames** [comment]
Can I clarify a couple of quick things:

1. Disruptor version - log4j source tree claims 3.4.2 - is that what's being run in the program having issues?
2. That consumer thread trace when the system is in a broken state:

"Log4j2-TF-1-AsyncLogger[sls-service]-1" #24 daemon prio=5 os_prio=0 cpu=7707684.49ms elapsed=23625.18s tid=0x00007fd28782a000 nid=0x45c2 runnable  [0x00007fc56066d000]
   java.lang.Thread.State: RUNNABLE
        at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:190)
        at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
        at java.lang.Thread.run(java.base@11.0.8/Thread.java:834)

Is this the only stack we see from the consumer thread? Have stacks have been repeatedly taken across multiple seconds? I would guess yes from the original comment, but I want to make sure before we look further.

Thanks,
James.

**Comment by carterkozak** [comment]
1. That is correct, the application in question was using disruptor `3.4.2` at the time of failure.
2. That is also correct, the batch event processor loop appears to spin indefinitely. Note the high cpu time compared to elapsed.

Thanks!

**Comment by grumpyjames** [comment]
Been thinking on this over the weekend. Haven't got any answers, but a few more questions.

It seems odd that we constantly pick up the read thread on the line from a closing brace. That seems like an unlikely outcome to me. What is that thread doing? Appending log messages to...somewhere? I'd expect to find the thread somewhere in client code rather than spinning in `BatchEventProcessor` - if repeated stack traces _always_ find us at that closing brace, that is interesting.

**Actions:**
  - We see the 'read' thread spinning, and the ring buffer appears to be constantly full. Can you tell us more about the metrics captured here? I'd be interested to know if, for instance, the read side's sequence was still going up. If there were charts of read sequence/write sequence/depth over time you could share, that would be terrific :-)
  - We should probably look into how a thread could get stuck at that brace. What's actually executing there? I'd start just by doing some experiments with a debugger attached.
  - The other possibility is that our jvm implementation is not telling us the whole truth. My knowledge here is out of date; I will attempt to remedy this...

If I trust that stack trace, this has the _feeling_ of a resource bottleneck hurting the read thread in some way more than a memory visibility issue. If we'd found the read thread blocking on something near the sequence classes, I'd be more tempted towards the latter.

On that note - is it worth attempting to create a test program that simulates the usage of this ringbuffer, potentially with a smaller buffer/higher input rate to see if this is reproducible? That might give us a bit more hope :-)

**Comment by carterkozak** [comment]
> I'd expect to find the thread somewhere in client code rather than spinning in BatchEventProcessor - if repeated stack traces always find us at that closing brace, that is interesting.

I don't have many samples unfortunately, however I am certain there is no progress on the queue.

> We see the 'read' thread spinning, and the ring buffer appears to be constantly full. Can you tell us more about the metrics captured here? I'd be interested to know if, for instance, the read side's sequence was still going up. If there were charts of read sequence/write sequence/depth over time you could share, that would be terrific :-)

We only record the remainingCapacity, not the read sequence/write sequence/depth. Are these reported in some way that I could begin collecting them for future occurrences? We also record metrics (increment a counter) each time an event is about to be handled, however there are no interactions after the queue begins to grow (remainingCapacity slowly drops after jitter within a few elements of the buffer size).

> We should probably look into how a thread could get stuck at that brace. What's actually executing there? I'd start just by doing some experiments with a debugger attached.

I don't think we're stuck in the brace, rather spinning over the loop without making progress. I haven't been able to reproduce the issue, and it's exceedingly rare. It happened again today for the first time since I reported this issue, unfortunately the server was  restarted before I could collect any diagnostics.

> is it worth attempting to create a test program that simulates the usage of this ringbuffer, potentially with a smaller buffer/higher input rate to see if this is reproducible? That might give us a bit more hope :-)

I can try this again, and leave it running for a few weeks on a cloud instance to see if I can force a reproducer. My attempts thus far have been unsuccessful.

**Comment by grumpyjames** [comment]
Going to write a couple of responses, one wordy, and one codey. This is the wordy one.

> I don't have many samples unfortunately, however I am certain there is no progress on the queue.

I am not yet convinced :-). It's possible for remaining capacity to remain zero and progress still to be happening. What does the buffer reading thread do with those messages? Are you also seeing that the effects of that thread are not happening? That would help to convince me that we were 'wedged' rather than 'hosed'. 

Offhand thought: Are we confident there are no failure modes for that read thread where it will attempt to log (leading to writing to its own input)?

<blockquote>
  A hosed system is processing data as fast as it can, but the input rate exceeds its maximum throughput, so its input ring buffer gradually fills up. 
  A wedged system has encountered some sort of problem to cause its throughput to be hugely reduced, or to stop altogether.</blockquote>

> We only record the remainingCapacity, not the read sequence/write sequence/depth. Are these reported in some way that I could begin collecting them for future occurrences? 

One extra metric that might also be worth tracking is the reading `BatchEventProcessor`'s `Sequence`. 

One **big** caveat here: That sequence only gets updated at the end of a batch (successful or otherwise). I see that log4j has attempted to callback to `BEP` every fifty items though - that might be enough to see if the reader is making progress.

> We also record metrics (increment a counter) each time an event is about to be handled, however there are no interactions after the queue begins to grow (remainingCapacity slowly drops after jitter within a few elements of the buffer size).

Can you point me to this code in the log4j source?

> I can try this again, and leave it running for a few weeks on a cloud instance to see if I can force a reproducer. My attempts thus far have been unsuccessful.

Good luck!

**Comment by grumpyjames** [comment]
This is the codey one. I want to talk about `BatchEventProcessor::processEvents`, which I will now reproduce here in full:

```
private void processEvents()
{
    T event = null;
// This is a local cache of sequence to avoid the costs of volatile write during a batch
    long nextSequence = sequence.get() + 1L;

    while (true)
    {
        try
        {
            final long availableSequence = sequenceBarrier.waitFor(nextSequence);
            if (batchStartAware != null)
            {
// This is another potential monitoring point here
                batchStartAware.onBatchStart(availableSequence - nextSequence + 1);
            }

// This could be a huge batch - event handlers can implement SequenceReportingEventHandler to call back with sequence numbers mid-batch
// I believe log4j does this every 50 messages in its RingBufferLogEventHandler::notifyCallback method
            while (nextSequence <= availableSequence)
            {
                event = dataProvider.get(nextSequence);
                eventHandler.onEvent(event, nextSequence, nextSequence == availableSequence);
                nextSequence++;
            }

// If the event handler completes successfully, we'll call the sequence to let upstream writers now some slots are free
            sequence.set(availableSequence);
        }
        catch (final TimeoutException e)
        {
// This should only happen when using wait strategies that can time out
// We could implement TimeoutHandler, increment a counter and monitor that in addition?
            notifyTimeout(sequence.get());
        }
        catch (final AlertException ex)
        {
// This is effectively a shutdown condition. We definitely haven't tried to stop this BatchEventProcessor, right? :-)
            if (running.get() != RUNNING)
            {
                break;
            }
        }
        catch (final Throwable ex)
        {
// Another possibility - if an exception handler is installed, could it be throwing?
            exceptionHandler.handleEventException(ex, nextSequence, event);
// If it isn't, we're going to update the sequence and make progress in this branch too.
            sequence.set(nextSequence);
            nextSequence++;
        }
    }
}
```

**Comment by grumpyjames** [comment]
It's all gone quiet here - were those last responses of mine just way too long?

Have you had any success reproducing the issue?

**Comment by carterkozak** [comment]
Hi @grumpyjames,

I'm very sorry for the radio silence, it has been both an unusual and busy few months.

I didn't have any luck reproducing the issue on a replica of the original environment, given the impact of failures I updated the logging configuration to fall back from full-ringbuffer to log on the calling thread (in the degraded case logging operates without any sort of asynchronous appenders, like a standard log4j/logback configuration) as opposed to blocking to enqueue events onto the ringbuffer. I haven't seen the bug occur since making this change, so it's possible it's either avoided, or able to heal when we never use the blocking enqueue method.

> I am not yet convinced :-). It's possible for remaining capacity to remain zero and progress still to be happening. What does the buffer reading thread do with those messages? Are you also seeing that the effects of that thread are not happening? That would help to convince me that we were 'wedged' rather than 'hosed'.

The consumer updates a codahale meter each time an event is received and pushed into the logging framework. When the deadlock occurred, these meters stopped incrementing.
Unfortunately the metrics I'm using are not part of the log4j framework, but an internal component that we use via log4j. Based on the stack trace location within the degraded state, we know the code is not blocked doing any logging appender work, and is in the disruptor loop.

> Offhand thought: Are we confident there are no failure modes for that read thread where it will attempt to log (leading to writing to its own input)?

We have put a great deal of work into ensuring that logging from the background thread will result in synchronous logging (direct access to the underlying appender resource instead of adding data to the ringbuffer). My understanding is if that did happen, we would see the background thread blocked attempting to enqueue into its own queue, and that's not what our telemetry indicates.

> One extra metric that might also be worth tracking is the reading BatchEventProcessor's Sequence.
> One big caveat here: That sequence only gets updated at the end of a batch (successful or otherwise). I see that log4j has attempted to callback to BEP every fifty items though - that might be enough to see if the reader is making progress.

I haven't wired this up, it's a little more complicated than one might hope due to the ringbuffer being encapsulated within the log4j-core codebase, while most of the metrics are wired externally.

>> We also record metrics (increment a counter) each time an event is about to be handled, however there are no interactions after the queue begins to grow (remainingCapacity slowly drops after jitter within a few elements of the buffer size).

> Can you point me to this code in the log4j source?

Unfortunately these metrics are created in an appender implementation we use internally, source is not available. We use a single appender which updates the metric each time it handles an event.

**Comment by grumpyjames** [comment]
> Hi @grumpyjames,
> 
> I'm very sorry for the radio silence, it has been both an unusual and busy few months.
>

No worries at all - thank you for getting back to me with an update; please accept my apologies for not noticing it for nearly a week!

If I can summarise your response, it looks like you have an alternative approach which avoids the original issue, and we're still not hugely the wiser about how to trigger the original problem. I also suspect that at this point, a reproducer for this is probably not going to appear. It feels like it may be time for the issue to go quietly into the night of "cannot reproduce", perhaps to be revived at a later date when I am least expecting it.

Let me know if that's not the case, and I'll arrange a stay of execution.

**Comment by rrossiersf** [comment]
Hello,
I am having the exact same issue. Very hard to reproduce and it is happening rarely (every 6 months) , but when it does it is a big problem for me. Publisher is reporting full buffer , but on the other side the reader is happily spinning not picking any new events. I am also into a multi producers , single consumer scenario. Not sure if it is going to help solve this problem but it is a problem that shows up on another system.

"pool-2-thread-17" #56 prio=5 os_prio=0 tid=0x000055d4a02d4800 nid=0x3e2a1 waiting on condition [0x00007f72dcad6000]
:    java.lang.Thread.State: TIMED_WAITING (parking)
: 	at sun.misc.Unsafe.park(Native Method)
: 	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:338)
: 	at com.lmax.disruptor.SleepingWaitStrategy.applyWaitMethod(SleepingWaitStrategy.java:92)
: 	at com.lmax.disruptor.SleepingWaitStrategy.waitFor(SleepingWaitStrategy.java:65)
: 	at com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:56)
: 	at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:159)
: 	at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
: 	at java.lang.Thread.run(Thread.java:748)
